File: 145-binary-tree-postorder-traversal/145-binary-tree-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */


/**

Approach 1 - O(n) Tc & O(height) SC

class Solution {
private:
    vector<int> postorder;
public:
    vector<int> postorderTraversal(TreeNode* root) {
        if(!root) return postorder;
        postorderTraversal(root->left);
        postorderTraversal(root->right);
        postorder.push_back(root->val);
        return postorder;
    }
};

Approach 2 - Morris traversal

*/

class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> postorder;
        if(!root) return postorder;
        TreeNode* current = root;
        while(current) {
            if(current->right) {
                TreeNode* rightSubTree = current->right;
                while(rightSubTree->left and rightSubTree->left != current)
                    rightSubTree = rightSubTree->left;
                if(!rightSubTree->left) {
                    postorder.push_back(current->val);
                    rightSubTree->left = current;
                    current = current->right;
                } else {
                    rightSubTree->left = nullptr;
                    current = current->left;
                }
            } else {
                postorder.push_back(current->val);
                current = current->left;
            }
        }
        reverse(postorder.begin(), postorder.end());
        return postorder;
    }
};

File: 145-binary-tree-postorder-traversal/145-binary-tree-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> postorder = new ArrayList<>();
        if(root == null) return postorder;
        TreeNode current = root;
        while(current != null) {
            if(current.right != null) {
                TreeNode rightSubTree = current.right;
                while(rightSubTree.left != null && rightSubTree.left != current)
                    rightSubTree = rightSubTree.left;
                if(rightSubTree.left == null) {
                    postorder.add(current.val);
                    rightSubTree.left = current;
                    current = current.right;
                } else {
                    rightSubTree.left = null;
                    current = current.left;
                }
            } else {
                postorder.add(current.val);
                current = current.left;
            }
        }
        Collections.reverse(postorder);
        return postorder;
    }
}